### ISSUE TITLE
Non-nullable field in a callback becomes nullable in the generated mock

### ISSUE THREAD

**Issue by miguelcmedeiros** [dart-lang/mockito#417]
When generating a mock for the following class, it makes the 2nd field in action nullable:
```
class SomeClass {
  void doSomething<T>(
    String? j,
    T Function(Map<String, dynamic>) action,
  ) {}
}

@GenerateMocks([SomeClass])
```

It generates:
```
/// A class which mocks [SomeClass].
///
/// See the documentation for Mockito's code generation for more information.
class MockSomeClass extends _i1.Mock implements _i10.SomeClass {
  MockSomeClass() {
    _i1.throwOnMissingStub(this);
  }

  @override
  void doSomething<T>(String? j, T Function(Map<String, dynamic>)? action) =>
      super.noSuchMethod(Invocation.method(#doSomething, [j, action]),
          returnValueForMissingStub: null);
}
```

`action` became nullable in the generated mock.

I'm using Mockito 5.0.9.

**Comment by miguelcmedeiros** [comment]
In case a it helps to have a real use-case, I found this while trying to gen the mock of `StreamChatClient` from https://pub.dev/packages/stream_chat/versions/2.0.0-nullsafety.2.

**Comment by srawlins** [comment]
This is the idea. All parameters must be made nullable so that they can accept argument matchers like `any`, `anyNamed`, etc.

**Comment by miguelcmedeiros** [comment]
In the case of `StreamChatClient` the problem was happening with `decode` method ([link](https://github.com/GetStream/stream-chat-flutter/blob/2.0.0-nullsafety.4/packages/stream_chat/lib/src/client.dart#L922)), which looks like:
```
T decode<T>(String? j, DecoderFunction<T> decoderFunction) {
  ...
}
```

Since the return type is a non-nullable generic type, I followed the [instructions](https://github.com/dart-lang/mockito/blob/master/NULL_SAFETY_README.md#fallback-generators) to define a fallback generator for `decode` and used something like:
```
T decodeShim<T>(String? j, DecoderFunction<T> decoderFunction) {
  throw Exception();
}

@GenerateMocks(
  [],
  customMocks: [
    MockSpec<StreamChatClient>(
      fallbackGenerators: {#decode: decodeShim},
    ),
  ],
)
```

I guess the problem was that I made the signature of `decodeShim` the same as `decode` (`decoderFunction` define as a non-nullable type). After changing the type in `decodeShim` to nullable, there were no more errors and the test ran without errors.

The issue is solved for me, but I wonder if either this can be handled by the code generator or if it should be mentioned in the documentation. I kind of learned my lesson with this one, but I can image others stumble on this.
